### ISSUE TITLE
Protocol sniffing causes excessive memory usage

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#16869]
When protocol sniffing is enabled, Envoy will use very high memory usage.

Some numbers:

* With 20 services, memory usage went from 40mb->80mb
* With 300 services, memory usage went from 60mb to 800mb
* With mixer disabled and 300 services, memory was around 400mb

These numbers are all taken from an Envoy receiving no traffic. Envoy will have high allocated as well, not just high heap usage (as seen in previous cases where it would use a lot of memory at startup then free it).

When switching from `http` to `auto`, we will no longer use a shared listener/HCM. So previously, if we had 100 HTTP services on port 80, we had 1 listener and 100 routes. Now, we will have 100 listeners, with 100 HCMs, and 100 routes. 

Envoy may internally also store a copy of the routes in the HCM as well, so that would result in 100*100 route configs (this is a theory only)

cc @rlenglet @rshriram  @yxue @lambdai @costinm

**Comment by rshriram** [comment]
This is 300 services without a Sidecar resource or with a Sidecar ?

**Comment by rshriram** [comment]
> When switching from http to auto, we will no longer use a shared listener/HCM. So previously, if we had 100 HTTP services on port 80, we had 1 listener and 100 routes. Now, we will have 100 listeners, with 100 HCMs, and 100 routes.

This is not the right way to view. We are not converting a pre-declared http port into auto. We are using auto if and only if the user did not provide a port name. 
So, If a user had 100 http services on port 80, all named as http, then we will only have one listener. If a user had 100 services on port 80, with no port names, then we will have 100 TCP listeners with one filterchain (without sniffing), and 100 listeners with two filterchains (with sniffing).

From that point of view, I find this stat not very convincing. If the argument here is that a very large cluster with 300+ services without any named ports is going to incur a much larger sidecar memory usage [*on the very first Istio usage attempt*], then my argument would be that this user has the option of only getting TCP telemetry without sniffing, or TCP and HTTP telemetry like Linkerd with sniffing. Beyond the first point of use, people can name ports and optimize the memory consumption. Sniffing is not about the end state where someone has thoroughly tested Istio and has rolled it into production after fixing up all the issues. Its about the first experience/usage where someone tries this out in a dev cluster and things work out of the box.

**Comment by howardjohn** [comment]
I have 300 services the entire time. in case 1 I use sidecar to only import 20, in case 2 I do not use sidecar and therefore get 300 service config

**Comment by howardjohn** [comment]
> This is not the right way to view. We are not converting a pre-declared http port into auto. We are using auto if and only if the user did not provide a port name.

If a user has unnamed ports right now they get tcp and with this feature they will get OOM. I think there are many services around that do not have named ports and they are ok with that because they are using tcp, and `foo` and `tcp` meant the same thing in 1.2.

**Comment by rshriram** [comment]
> if a user has unnamed ports right now they get tcp and with this feature they will get OOM.

This needs to be qualified. If this is a large scale user (one who has 300 services), then he/she will certainly have to disable protocol sniffing in Pilot before deploying/upgrading Istio. But they already do such customizations given that we have an armada of env vars.

If this is a small scale or average user with 20-50 services, I do not think this will result in OOM after Istio upgrade, given that these users would have already named most of their ports.

>  I think there are many services around that do not have named ports 
This is a stat that needs to be backed by data in terms of number of people with "many un named ports".

My point here is that the contrived experiment in this issue is effectively concluding that protocol sniffing in Envoy is pointless as it leads to memory bloat, and that we should dump all efforts to simplify the onerous burden on end users to name ports properly on first attempt - when other projects like Linkerd have effectively used the sniffing logic to dramatically simplify initial user onboarding. If this is the intended conclusion, then we should go back to the drawing board to see if we can really simplify Istio/Envoy user experience or if we should give up due to constant fear of scale issues.

**Comment by howardjohn** [comment]
I don't think we can consider it a contrived experiment. I did not even intentional test this, I just deployed a pod into my cluster and it was immediately oomkilled.

No one came to the conclusion that protocol sniffing is pointless or that we should give up - the only conclusion was that our current implementation is not efficient enough to be viable for users.

Would you suggest we ship protocol sniffing but tell users "if you use protocol sniffing your pods will use 10x memory"? I think that is an even worse UX than what we have today.


We found this issue at 4pm today and we release in ~24 hours. turning it off by default now does not mean we cannot fix this and turn it on later - Yan is working right now to see how we can reduce the memory impact.

**Comment by kyessenov** [comment]
Is it true that having 300 listeners in envoy causes 400MB of use? If that's the case, then that's the real problem here. Envoy needs to be able to represent 300 TCP routes without incurring gigabytes of RAM via compressed/generic/templatized config.

**Comment by rshriram** [comment]
@howardjohn the contrived piece referred to the 300 services use case and the notion of auto on all of these services. The memory consumption issue was very well known before hand (doubling filter chains results in more stats and more memory). What is surprising as @kyessenov points out is the 400Mb usage. That seems dramatic, especially since you said

> With 300 services, memory usage went from 60mb to 800mb

If in the above case, you had 300 TCP listeners (with one filter chain) and envoy was consuming 60Mb of memory, then how does it grow by an order of magnitude to 800Mb when you simply add another filter chain match in each listener? And they are all using the same RDS route (for a port) today btw. They are not using separate RDS routes [thats a todo in code].
So do please check your setup. Ensure you did not turn on something else like full stats in every proxy.

Also, if make sure your base line 60Mb is correct with 300 TCP listeners with one filter chain

**Comment by howardjohn** [comment]
Kuat it's not the tcp routes that are the problem - if I create 300 tcp services the memory usage was small. It's only when we create 300 http connection managers that usage is so high. Looking at heap profile (I should have attached here, will do tomorrow) shows most memory is in the routes, which is what led me to believe there is a HCM*Routes scaling problem, even though each HCM points to the same routes, so in theory it could be HCM+Routes scaling.

**Comment by kyessenov** [comment]
@howardjohn This is equivalent to having 300 HTTP services each with a different port and no sniffing, right? Then again, this is the true issue, 300 http connection managers is not that many and should not consume a gigabyte. I agree with you that we have little wiggle room here to safely enable sniffing.

**Comment by howardjohn** [comment]
I don't think it is the same. If you have 300 http services with different ports you get 300 HCM and each points to a route config with 1 route. In our case each one point to a route config with 300 routes. In terms of config_dump size it's the same, but it seems Envoy internal representation it is not the same .

This is just the running theory though.. I am not familiar with Envoy internals

**Comment by kyessenov** [comment]
Ah, so each HCM points to the same route config with 300 routes. So logically it's 300**2 routes, but we're not sure if Envoy compresses them correctly.

**Comment by howardjohn** [comment]
Yep that's my guess - we will need to confirm this. Yan is looking into if we can just make each one point to a single route (seems like it should work)

**Comment by kyessenov** [comment]
I suspect there'll be 300**2 copies of the proto in memory for routes. It's because subscriptions have to unwrap any per each HCM, so the routes are not shared between HCMs.

**Comment by rshriram** [comment]
Yes this is a known issue and I left a TODO for @yxue a while ago in the RDS code saying that the RDS output will have to be trimmed. I didn't think much about envoy level memory usage as I thought envoy would be storing only one copy of the RDS response, but apparently not.

**Comment by rshriram** [comment]
that hack is easy though..  just attach the hostname to the route name (like 80.foo.svc.cluster.local), (its a nasty hack) and strip out the generated routes to return the route for that host alone.

**Comment by howardjohn** [comment]
Well Yan is working on that now, so hopefully it's fairly simple and we can
have something soon. We are building the release tomorrow but not planning
to ship until Tuesday afaik. If we could fix by tomorrow or Monday then I
suspect we could convince the release managers to delay the release by a
day or 2, but I don't think they will accept indefinitely delaying the
release until this is fixed.

**Comment by costinm** [comment]
The plan discussed with the rel manager was:
- turn off sniffing on outbound for tomorrow release, leave it for inbound
only ( so we get metrics when we capture all )
- attempt to generate in-line routes with one host only - if it works we
can add this to 1.3.1 ( or even 1.3.0 if it is delayed for independent
reasons ).

My concern is longer term: we still want to switch to filter chains, and it
is not clear how the sniffing will work there.
Right now on 15000 we have a listener that forwards to the virtual
listeners, and each virtual listener has control over sniffing or not.
With filter chains it doesn't seem to be an option, some changes will
likely need to be made.

I don't think this is an uncommon example - we do support 300 services
without sidecar in 1.2, and we can expect users with as many
services or more. Yes, they could turn on sidecar and isolate - but that
hasn't been needed for small deployments.

The fundamental requirement I think is to not break users who just upgrade
Istio. Even with smaller number of services there
is still a significant grow and may cause sidecar OOM.

My preference is to have it off by default for outbound, if
one-route-per-sniffed-listener reduces the pressure.

**Comment by howardjohn** [comment]
Testing with Mixer and no Sidecar, 300 services still.

With rc1: 775Mi
With yans patch: 120Mi
With yans patch + 1.2 proxy (meaning no protocol sniffing, everything is TCP): 90mi
With yans patch, all ports are named, 1.2 proxy: TBD
With yans patch, all ports are named, 1.3 proxy: TBD

**Comment by rlenglet** [comment]
All fix PRs have been merged into `release-1.3` and will make it into `1.3.0-rc.2`.

**Comment by duderino** [comment]
+1!

**Comment by howardjohn** [comment]
Running rc2:

* Pilot uses 4 CPUs with protocol sniffing (2 replicas), 1 CPU without (single replica).
* Pilot uses 750mb heap, 450mb without sniffing. Note this is skewed by # of replicas.
* In a Sidecar scoped ns, Envoy memory is 40mb with sniffing, 30mb without
* In non Sidecar scoped ns, Envoy memory is 170mb with sniffing, 90mb without

With sniffing, building the routes accounts for 78% of pilot allocations and 53% of CPU.
Without sniffing, the routes use 8% of allocations and 6% of CPU.

So we are losing all Pilot performance gains from 1.3 and 33-100% more memory in Envoy. Should be fixable soon though (the pilot stuff, Envoy side probably won't change)

**Comment by sdake** [comment]
> Running rc2:
> 
> * Pilot uses 4 CPUs with protocol sniffing (2 replicas), 1 CPU without (single replica).
> * Pilot uses 750mb heap, 450mb without sniffing. Note this is skewed by # of replicas.
> * In a Sidecar scoped ns, Envoy memory is 40mb with sniffing, 30mb without
> * In non Sidecar scoped ns, Envoy memory is 170mb with sniffing, 90mb without
> 
> With sniffing, building the routes accounts for 78% of pilot allocations and 53% of CPU.
> Without sniffing, the routes use 8% of allocations and 6% of CPU.
> 
> So we are losing all Pilot performance gains from 1.3 and 33-100% more memory in Envoy. Should be fixable soon though (the pilot stuff, Envoy side probably won't change)
> 
> When I switched just half of the services to `auto` and half named ports, pilot uses 2 CPUs, so it seems to scale linearly with number of unnamed ports as expected, which is good as it means most users today will not see an impact at upgrade (UNLESS they purposely don't label services as `tcp` since they know `foo` means `tcp`)

@howardjohn I know all original master PRs went through an extremely thorough review on these set of changes.  Did you see any potential problems in regards to crashes or regressions?

Cheers
-steve

**Comment by howardjohn** [comment]
@sdake we know exactly what the cause of the performance regression is, so I don't think there is a risk of crashes, unless there was OOM from increased memory usage)

Cause of envoy memory is that we are generating a http config for every service rather than sharing a single config

Cause of pilot CPU usage is O(n^2) route generation, when it was/could be O(n)

**Comment by sdake** [comment]
@howardjohn 

Admittedly not a Pilot expert and didn't review the PRs in question:

> @sdake we know exactly what the cause of the performance regression is, so I don't think there is a risk of crashes, unless there was OOM from increased memory usage)
> 
> Cause of envoy memory is that we are generating a http config for every service rather than sharing a single config
> 

Sharing of in-memory config in multithreaded applications can cause problems.  100% sure there will be no deadlocks?  Would we risk a space flight on this default?

> Cause of pilot CPU usage is O(n^2) route generation, when it was/could be O(n)

This seems less of a risk, as the route generation code path is well-trodden.

In my mind, it is a simple tactical equation for Istio to measure.  Is the risk of a DOA 1.3.0 release worth the desire to enable this (new) feature by default or packing on a few more pounds of testing?

I understand the usability benefits: naming ports is really really annoying, and I get the competitive pressure as well.  Does Istio leadership understand the potential for a DOA 1.3.0 vs the tradeoffs of not releasing on film day?

@duderino any thought on the above?

Cheers
-steve

**Comment by howardjohn** [comment]
> Sharing of in-memory config in multithreaded applications can cause problems. 100% sure there will be no deadlocks? Would we risk a space flight on this default?

The problem is actually that it doesn't share memory it copies it so it used a ton of memory. Anyhow with rc2 we no longer are sharing config anyways, so I'm not to concerned on this front. There is of course always some risk, but in not too concerned about Envoy deadlocks - it is pretty well battle tested and we are using standard features.

There's plenty of risks in shipping this, but I think deadlocks are not a concern.

**Comment by rshriram** [comment]
I can take a crack at the rds optimization before release but would need a hand to finish it. Will report by Monday

**Comment by rshriram** [comment]
see #16921
